docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,488 @@
+
+Another option doesn't create separate configuration for each client, but uses just Mesh scoped `MeshLoadBalancingStrategy` to configure egress. That reduces possibilites of configuring Egress but can be changed in the future.
+
+The most favorable approach appears to be defining routing on the client side. Instead of creating separate cluster sets for each client on the Egress, we create an 'aggregate_cluster' for the client:
+
+Aggregate_cluster:
+
+- backend_in_zone
+- backend_zone1
+- backend_zone2
+
+Each entry represents the Envoy cluster name that the client is interested in communicating with, in order of priority. This arrangement enables each cluster to set a separate 'SNI' for each zone, facilitating matching and routing on the Egress side.
+
+For example:
+
+```yaml
+backend_zone1 = SNI backend{mesh=default,kuma.io/zone=zone1}
+backend_zone2 = SNI backend{mesh=default,kuma.io/zone=zone2}
+```
+
+In addition, it is important to configure the Egress correctly. For each `SNI`, we need to create a filter chain match that sets the required metadata for selecting endpoints:
+
+```yaml
+- filterChainMatch:
+    serverNames:
+    - backend{mesh=default,kuma.io/zone=zone1}
+    transportProtocol: tls
+  filters:
+  - name: envoy.filters.network.tcp_proxy
+    typedConfig:
+      '@type': type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
+      cluster: default:backend
+      metadataMatch:
+        filterMetadata:
+          envoy.lb:
+            kuma.io/zone: zone1
+      statPrefix: default:backend_zone1
+```
+Each dataplane has metadata, and it's important to ensure that `kuma.io/zone` is available. If it's not present, add it to the endpoints.
+
+On the ingress side, you only need to adjust the `SNI` to include its corresponding `kuma.io/zone`.
+
+However, it's important to note that this step may not be within the scope of the initial implementation. We are going to support initially: **Support only Mesh scope for Egress**.
+
+##### Ingress
+We don't configure ingresses.
+
+#### ClusterLoadBalancing cache
+Currently, while retrieving endpoints in the plugin, we are getting a pointer to the object in the cache that is shared between dataplanes. This won't work with localityAwareness load balancing because each dataplane might have a different subset of endpoints with different priorities.
+
+We are going to implement this without the cache and verify performance.
+
+Steps:
+* Run `mesh-perf` test to measure impact of not using CLA cache
+* If results are good we'll remove the CLA cache
+* Otherwise rewrite CLA cache to keep the different clusters depending on the dataplane tags
+
+##### Test result
+We have ran a test of 1000 services with 2 instances each in standalone mode. Each service had reachable services to 3 services defined.
+
+![With CLA Cache](assets/031/xds_generation_with_cache.png)
```

Comment:
nit: when viewing the MADR, there are no labels indicating which graph is with and which is without

Reply:
good point, changed